components/style/context.rs

```diff
@@ -76,25 +75,25 @@ pub struct SharedStyleContext {
     pub quirks_mode: QuirksMode,
 }
 
-pub struct ThreadLocalStyleContext {
-    pub style_sharing_candidate_cache: RefCell<StyleSharingCandidateCache>,
+pub struct ThreadLocalStyleContext<E: TElement> {
+    pub style_sharing_candidate_cache: StyleSharingCandidateCache<E>,
```

Comment:
From IRC:

Are you proposing adding an additional lifetime parameter to StyleSharingCandidate beyond the one implicit in E? The lifetime in E is the thing that prevents the cache entries from outliving the Element borrows. I don't think we could annotate DomTraversalContext with the scope lifetime, because the scope is created during the traversal.

In general, having a lifetime on DomTraversalContext was a PITA, and was the reason for the nasty transmute that I removed in #14610.